### PR TITLE
Hub 1044

### PR DESCRIPTION
**Pull request by Paul-Hess**

### CODE REVIEW

**Comment by lkemmerer** [comment]
Let me know if you want to discuss the comments. They're a bit rambling, much like the hub code itself.

File: src/main/java/com/flightstats/hub/dao/aws/S3Config.java [hunk 1]
```diff
@@ -86,17 +87,20 @@ protected Scheduler scheduler() {
...
-            maxItemsEnforcer.updateMaxItems(configurations);
+            Stream.of(s3Properties.getBucketName(), s3Properties.getDisasterRecoveryBucketName())
+                    .forEach(bucketName -> {
+                        updateTtlDays(bucketName);
+                        maxItemsEnforcer.updateMaxItems(configurations, bucketName);
+                    });
         }
 
-        private void updateTtlDays() {
+        private void updateTtlDays(String bucketName) {
```
[the remark this answers is not in the document]
Comment (Paul-Hess) [reply src/main/java/com/flightstats/hub/dao/aws/S3Config.java:103]: Updated

File: src/main/java/com/flightstats/hub/dao/aws/S3Config.java [hunk 2]
```diff
@@ -86,17 +87,20 @@ protected Scheduler scheduler() {
...
         }
 
         @Override
         public void takeLeadership(Leadership leadership) {
-            updateTtlDays();
-            maxItemsEnforcer.updateMaxItems(configurations);
+            Stream.of(s3Properties.getBucketName(), s3Properties.getDisasterRecoveryBucketName())
+                    .forEach(bucketName -> {
+                        updateTtlDays(bucketName);
+                        maxItemsEnforcer.updateMaxItems(configurations, bucketName);
```
[the remark this answers is not in the document]
Comment (Paul-Hess) [reply src/main/java/com/flightstats/hub/dao/aws/S3Config.java:99]: Updated

File: src/main/java/com/flightstats/hub/dao/aws/MaxItemsEnforcer.java [hunk 3]
```diff
@@ -66,7 +66,24 @@ private void updateMaxItems(ChannelConfig config) {
 
     }
 
-    private void updateMaxItems(ChannelConfig config, ContentKey latest) {
+
+    private void updateMaxItems(ChannelConfig config, String bucketName) {
+        String name = config.getDisplayName();
+        log.debug("updating max items for channel {}", name);
+        ActiveTraces.start("S3Config.updateMaxItems", name);
+        Optional<ContentKey> optional = contentRetriever.getLatest(name, false);
```
[the remark this answers is not in the document]
Comment (Paul-Hess) [reply src/main/java/com/flightstats/hub/dao/aws/MaxItemsEnforcer.java]: done

File: src/main/java/com/flightstats/hub/dao/aws/S3Config.java [hunk 4]
```diff
@@ -96,35 +97,35 @@ public void takeLeadership(Leadership leadership) {
             maxItemsEnforcer.updateMaxItems(configurations);
         }
 
+        private void updateRulesConfigForBucket(List<BucketLifecycleConfiguration.Rule> rules, String bucket) {
+            rules.addAll(getNonHubBucketLifecycleRules(bucket));
```
Comment (lkemmerer) [inline src/main/java/com/flightstats/hub/dao/aws/S3Config.java:101]: Is this line going to cause `rules` to build up non-hub-created rules such that the second bucket that loops through will get both its non-hub rules AND the rules of the bucket(s) before it?

If so, the DR bucket is slowly going to build up copies of the normal bucket's static rules, e.g. 
1. run 1 will add x static rules from the hub (DR bucket: x static rules + hub rules)
2. run 2 will add x static rules from bucket + x static rules from run 1 (DR bucket: 2x static rules + hub rules)
3. run 3: DR bucket ends up with 3x static rules + hub rules

**Review by lkemmerer — APPROVED** [review]
That mutable `addAll` thing is my only concern. Approving with the understanding that that'll be looked into and fixed if necessary.

File: src/main/java/com/flightstats/hub/dao/aws/S3Config.java (hunk 4 above)
Comment (Paul-Hess) [reply src/main/java/com/flightstats/hub/dao/aws/S3Config.java:101]: good catch

Comment (Paul-Hess) [reply src/main/java/com/flightstats/hub/dao/aws/S3Config.java]: fixed

**Pull request merged**